tests/conftest.py

```diff
@@ -41,106 +57,52 @@ def driver() -> Iterator[webdriver.Chrome]:
         The web driver to use in Selenium tests.
     """
     driver = selenium_driver()
-    yield driver
-    driver.quit()
+    try:
+        yield driver
+    finally:
+        driver.quit()
 
 
 @pytest.fixture
-def responses() -> Iterable[aioresponses]:
-    """Create an aioresponses context manager.
+def non_mocked_hosts() -> List[str]:
+    """Disable pytest-httpx mocking for the test application."""
+    return [TEST_HOSTNAME, "localhost"]
+
 
-    This can be used to mock responses to calls in an `aiohttp.ClientSession`.
+@pytest.fixture
+def selenium_server_url(tmp_path: Path) -> Iterable[str]:
+    """Start a server for Selenium tests.
+
+    The server will be automatically stopped at the end of the test.
 
     Returns
     -------
-    mock : `aioresponses.aioresponses`
-        The mock object with which URLs and callbacks can be registered.
+    server_url : `str`
+        The URL to use to contact that server.
     """
-    with aioresponses(passthrough=["http://127.0.0.1"]) as mock:
-        yield mock
+    settings_path = build_settings(tmp_path, "selenium")
+    with run_app(tmp_path, settings_path) as server_url:
+        yield server_url
 
 
 @pytest.fixture
-def create_test_setup(
-    tmp_path: Path,
-    responses: aioresponses,
-    aiohttp_client: Callable[[web.Application], Awaitable[TestClient]],
-) -> SetupTestCallable:
-    """Create a test setup.
-
-    Parameters
-    ----------
-    tmp_path : `pathlib.Path`
-        Path to a per-test temporary directory, injected as a fixture.
-    responses : `aioresponses.aioresponses`
-        Mock object for `aiohttp.ClientSession` call handling.
-    aiohttp_client : `typing.Callable`
-        Function creates an aiohttp test client.
+async def setup(
```

Comment:
Something that's common in other applications I've worked with is that the test client fixture is responsible for setting up the application instance. That way you don't need to have two fixtures for every test.

On the other hand, maybe this approach is more transparent and flexible? I know that for some apps I'd need to create multiple client fixtures to have different application set ups (like a client that has different roles assigned to it).

Since the `client` fixture already depends on the `setup` fixture, I wonder if a test _could_ actually just use the `client` fixture in cases where it doesn't use the `setup` directly (i.e., is the docstring below still correct?).

Reply:
What the docstring is getting at is that `SetupTest` is responsible for creating the configuration, and until the configuration is created, attempting to create a client wrapping the ASGI application will immediately fail because it will try to run the lifecycle manager on the ASGI app, which will trigger the startup event, which will try to read the configuration file, and that will fail because `/etc/gafaelfawr/gafaelfawr.yaml` doesn't exist.

I'm going to try to turn `SetupTest` into an async context manager so that it can do both its own setup and that of the client and eliminate one of the fixtures. Stay tuned....

Reply:
I think that worked! Good idea.